consensus/src/execution_ctx.rs

```diff
@@ -149,79 +153,108 @@ impl<'a, DB: Database, T: Operations + 'static> ExecutionCtx<'a, DB, T> {
-                    event = "vote for former candidate",
-                    step_topic = format!("{:?}", topic),
-                    hash = to_str(&candidate.header().hash),
-                    msg_step,
-                );
+    pub(crate) async fn try_cast_ratification_vote(
+        &self,
+        msg_iteration: u8,
+        validation: &ValidationResult,
+    ) {
+        let step = StepName::Ratification.to_step(msg_iteration);
 
-                // TODO: Verify
-            };
-        } else {
-            error!(event = "committee not found", step = self.step(), msg_step);
+        if let Some(committee) = self.iter_ctx.committees.get_committee(step) {
+            if self.am_member(committee) {
+                RatificationStep::<T, DB>::try_vote(
+                    &self.round_update,
+                    msg_iteration,
+                    validation,
+                    self.outbound.clone(),
+                )
+                .await;
+            }
         }
     }
 
     /// Process messages from past
     async fn process_past_events(&mut self, msg: Message) -> Option<Message> {
-        if msg.header.round != self.round_update.round {
+        if msg.header.round != self.round_update.round
+            || self.iteration < EMERGENCY_MODE_ITERATION_THRESHOLD
+        {
+            // Discard messages from past if current iteration is not considered
+            // an emergency iteration
             return None;
         }
 
+        self.on_emergency_mode(msg).await
+    }
+
+    /// Handles a consensus message in emergency mode
+    async fn on_emergency_mode(&mut self, msg: Message) -> Option<Message> {
         if let Err(e) = self.outbound.send(msg.clone()).await {
             error!("could not send msg due to {:?}", e);
         }
 
-        // Try to vote for candidate block from former iteration
+        // Try to cast validation vote for a candidate block from former
+        // iteration
         if let Payload::Candidate(p) = &msg.payload {
-            // TODO: Perform block header/ Certificate full verification
-            // To be addressed with another PR
-
-            self.vote_for_former_candidate(msg.header.get_step(), &p.candidate)
+            self.try_cast_validation_vote(msg.header.iteration, &p.candidate)
                 .await;
         }
 
-        // Collect message from a previous reduction step/iteration.
+        let msg_iteration = msg.header.iteration;
+
+        // Collect message from a previous iteration/step.
         if let Some(m) = self
             .iter_ctx
             .collect_past_event(&self.round_update, msg)
             .await
         {
-            if m.header.topic == Topics::Quorum {
-                debug!(
-                    event = "quorum",
-                    src = "prev_step",
-                    msg_step = m.header.get_step(),
-                    hash = to_str(&m.header.block_hash),
-                );
+            match &m.payload {
```

Comment:
Why don't we handle all cases (including Candidate) inside `collect_past_event`?

Reply:
IMO there is no reason to handle the remaining cases. What logic we need to impl for them?

Reply:
@fed-franz 
the message return by `collect_past_event` is the optional result message (and not the message you're collecting)

and btw the former iteration/step candidate block is handled with `try_cast_validation_vote`

Reply:
What I meant is to have `collect_past_event` both collect and handle the type of message and its result:
if msg is Candidate, it collects the message and calls `try_cast_validation_vote`; if msg is Validation, it collects the message and then handle Quorum; if msg is Ratification, it collects the message and handle ValidationResult by calling `try_cast_ratification_vote`.
Does that make sense? It would be cleaner IMO

Reply:
> What I meant is to have collect_past_event both collect and handle the type of message and its result
> ..
> It would be cleaner IMO

IMO this is not cleaner as the idea of method `collect_past_event` is to be a wrapper of all phase.collect_past_event calls and to perform single function